### PR TITLE
Revised request API

### PR DESCRIPTION
**Pull request by mtmk**
Revised request(many) API based on #75 
Also added start-up timer as per ADR-40. It's the max time to wait for the first message.

### CODE REVIEW

**Comment by mtmk** [comment]
Are we happy with the public interface i.e. `NatsSub` and `Request*Async`?

* I had doubts about the options being passed to `Request` but I can see the value of not littering the API with too many types.

* `RequestMany` returning async enumerator forces the developer to use `await foreach`. Returning a type with a channel e.g. `NatsSub` may be a better option.

**Comment by caleblloyd** [comment]
For the `Request` options - we need a way to pass Queue Group, Headers, and a Serializer when using Model types.  So the only thing missing from `PubOpts` is the ReplySubject which is defaulted.

For the `Reply` options of the `RequestAsync` methods we could cut those down, as many of the timeouts do not apply when waiting for a single reply.  Or we could document that they have no effect when waiting for a single reply.

For the `Reply` options of `RequestManyAsync` - this pretty much does need to be the full set of `NatsSubOpts`

I think that returning `NatsSub` is fine for the `RequestManyAsync` methods

File: src/NATS.Client.Core/NatsRequestManyAsyncExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    // Same as PublishAsync with the following changes
+    // - Response is IAsyncEnumerable<NatsMsg>
+    // - PubOpts is called requestOpts
+    // - add SubOpts replyOpts
+    //   - if replyOpts.Timeout == null then set to NatsOptions.RequestTimeout
+    public static async IAsyncEnumerable<NatsMsg> RequestManyAsync(
+        this NatsConnection nats,
+        string subject,
+        ReadOnlySequence<byte> payload = default,
+        NatsPubOpts? requestOpts = default, // Not used
```
Comment (caleblloyd) [inline src/NATS.Client.Core/NatsRequestManyAsyncExtensions.cs:19]: I think this could be used - Headers and Queue Group could be set here right?

Comment (mtmk) [reply src/NATS.Client.Core/NatsRequestManyAsyncExtensions.cs:19]: Yes, spot on! Thank you. I forgot about headers when publishing. Queue group doesn't apply to pub-opts. But a good question. Does it make sense to have a queue group on an inbox subscription? If it does, how should we go about it when our inbox is mux-ed and shared, as in this case?

**Comment by mtmk** [comment]
> For the `Request` options - we need a way to pass Queue Group and Headers when not using Model types. So the only thing missing from `NatsPubOpts` is the ReplySubject which is defaulted.

see comment above

> For the `Reply` options of the `RequestAsync` methods we could cut those down, as many of the timeouts do not apply when waiting for a single reply. Or we could document that they have no effect when waiting for a single reply.

I'm happy with documenting it. Changed my mind a little. Maybe it's better to keep Type footprint to a minimum. Also, in my opinion at this level of the API it's better to give clues that under the bonnet it is still a pub/sub powering the request reply pattern. We can certainly wrap this later with a much more opinionated/easy/lite API for less demanding applications.

> For the `Reply` options of `RequestManyAsync` - this pretty much does need to be the full set of `NatsSubOpts`

Agreed.

> I think that returning `NatsSub` is fine for the `RequestManyAsync` methods

I was thinking to wrap this PR up as a smaller step and discuss this on a some kind of 'internal redesign' PR. Happy to carry on here though if you like.

File: src/NATS.Client.Core/NatsRequestManyAsyncExtensions.cs (hunk 1 above)
Comment (caleblloyd) [reply src/NATS.Client.Core/NatsRequestManyAsyncExtensions.cs:19]: Oops - yes - queue group is on subscribe.  Good catch.  No, I don't think an Inbox Subscription really needs to support a Queue Group, but it's also not a big deal if it does

**Comment by caleblloyd** [comment]
> I was thinking to wrap this PR up as a smaller step and discuss this on a some kind of 'internal redesign' PR

That sounds good!

**Review by caleblloyd — APPROVED** [review]
LGTM

**Pull request merged**